### PR TITLE
feat(tooltip): add definition tooltip

### PR DESCRIPTION
**Pull request by aledavila**
## Overview

https://github.ibm.com/carbon/issues/issues/618
https://github.ibm.com/carbon/issues/issues/549#issuecomment-4988198 
https://github.com/carbon-design-system/carbon-components/issues/771

### Added

Added new tooltips with on click tooltip being the default.
Added definition tooltip
Added icon tooltip

### Removed

Removed on hover tooltip.
Removed simple tooltip

### Changed

Simple tooltip is just for defining the icon.
Default font weight is now normal. Bold is a modifier. 

## Notes

Tayler needs to sign of and look at styles.

### CODE REVIEW

**Comment by alisonjoseph** [comment]
@aledavila These don't look like they are related to this PR?  #618 #549

File: src/components/tooltip/_tooltip.scss [hunk 1]
```diff
@@ -140,13 +142,12 @@
     &:before {
       @include typescale('zeta');
       @include layer('overlay');
-      max-width: rem(240px);
-      padding: $spacing-lg;
+      max-width: rem(170px);
+      padding: rem(10px) $spacing-xs;
```
Comment (alisonjoseph) [inline src/components/tooltip/_tooltip.scss]: Might be a question for design but can `rem(10px)` be tweaked up or down to fit into our spacing scale?

**Comment by aledavila** [comment]
@alisonjoseph yes you are right sorry. those are the issue #s in IBM carbon repo no this repo

**Comment by aledavila** [comment]
Fixed the links

**Review by alisonjoseph — CHANGES_REQUESTED** [review]
Overall looks good 😄 

Looks like the triangle needs to move 1px down? Screen shot of a screen shot I zoomed in on. 
screen shot 2018-04-19 at 12 40 28 pm

Changing the functionality of the existing tooltip would be a breaking change. (from hover to click). Would maybe be better to add it as a new variant or v2, thoughts?

**Comment by aledavila** [comment]
Thanks I’ll look into the changes. As it stands how the changes were made it doesn’t break anything rather people who have the tooltip implemented would have to click on it now instead of hovering when they pull the new changes. But i don’t know how you guys prefer to handle that. 

Yea I’ll ask Tayler about the changes probably better to have the padding all spacing-xs.

**Comment by asudoh** [comment]
Agreed with @alisonjoseph that looks great overall 🎉 - Thanks @aledavila!

The only question I have is how to handle the change in behavior, which would need to involve designers (CC @carbon-design-system/design) - Should we introduce some way to keep the older behavior, or should we try something to ensure that the change in behavior does not come with a surprise to our consuming services...?

**Comment by aledavila** [comment]
@alisonjoseph  I changed the 1 pixel that was off. I'll ask Tayler about the other padding issue when she gets back.

File: src/components/tooltip/_tooltip.scss (hunk 1 above)
Comment (aagonzales) [reply src/components/tooltip/_tooltip.scss]: It should be 8px

**Comment by aagonzales** [comment]
> Changing the functionality of the existing tooltip would be a breaking change. (from hover to click). Would maybe be better to add it as a new variant or v2, thoughts?

I think making it a v2 and deprecating the v1 version with Carbon V10 would be fine. Consumers just won't ever automatically get those update, correct? They'll have to manually update it to v2?

**Comment by alisonjoseph** [comment]
correct @aagonzales. That's how we've been handling other components with breaking changes.

**Comment by aledavila** [comment]
Oh I see. Can someone explain how that process is or where to make it a V2 instead?

**Comment by alisonjoseph** [comment]
Looks like the process for the other components is just updating the classname, so `bx--tooltip-v2` and just make sure that the original tooltip doesn't change. You can see how the v2 input styles are handled in this PR for an example. https://github.com/carbon-design-system/carbon-components/pull/712

**Comment by aledavila** [comment]
@alisonjoseph I added the v2 now. I just need to ask Tayler about the padding on the simple tooltip. Everything else should be good.

File: src/components/tooltip-v2/tooltip-v2--simple.html [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<div class="bx--tooltip--simple-v2">
```
Comment (alisonjoseph) [inline src/components/tooltip-v2/tooltip-v2--simple.html]: Should this class be `bx--tooltip-v2--simple`?

File: src/components/tooltip-v2/migrate-to-7.x.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
+### HTML
```
Comment (asudoh) [inline src/components/tooltip-v2/migrate-to-7.x.md]: Given V2 is newly introduced component, does this migration guide still make sense? One thing that comes to my mind is V1 to V2 migration guide. Another approach may be simply keeping the README in-sync with what V2 does.

**Review by asudoh — COMMENTED** [review]
Hi @aledavila great to see you complete V2 split!

**Comment by tay-aitken** [comment]
@aledavila can you create a staging link so that I can inspect the element and see how everything looks. Thanks! This will help me make a decision about the padding.

**Comment by tay-aitken** [comment]
### Fixes
Please take a a look at the [specs](https://github.ibm.com/carbon/issues/issues/549#issuecomment-5477046) I posted previously to make sure the coded variation matches the designs and make the following changes below:

- [ ] Change default tooltip name to Interactive
- [x] Change simple name to Icon

**Interactive tooltip**
- [x] Using the incorrect icon. Let me know if you need an SVG of the right one/if it needs to be added to our icon list. The SVG can be found [here](https://ibm.box.com/s/2wew5q4lzrfocdzcsomcv6vxybmll6bb)
- [x] add in a link and button as seen [here](https://github.ibm.com/carbon/issues/issues/549#issuecomment-5477046)
- [x]  Tooltip should not close if i click inside the tooltip, should only close if I click outside of the tooltip area
- [x] background is not black, it should be `background-color: #272D33;` (and should be using our color tokens)
- [x] Tooltip icon should be darkening on hover by 10% not lightening (sync with TJ, his PR might already be taking care of this)
- [x] Tooltip label should not be changing color on hover, should remain #152935
- [x] Can't inspect the tooltip box itself, having trouble locating it in the code, will review colors, type sizes, padding, etc. once link and button are added in

**Definition Tooltip**
- [x] Please make the tooltip have the following content "Brief description of the dotted, underlined word above."
- [x]  Is there a reason the old CSS for the tooltip still is a part of this component?
- [x] Padding on all sides should be .05rem
- [x] Dotted underline should be #8897A2 (ui-04) on enabled
- [x]  on :hover it should be #3d70b2 (brand-01)
- [x] should be `font-size: 0.75rem;` not 0.875rem
- [x] Tooltip icon should be darkening on hover by 10% not lightening (sync with TJ, his PR might already be taking care of this)
- [x] Double check that the tooltip is only 170px wide, perhaps add in a max-width
- [x] Change label to Definition label, instead of Tooltip label

**Icon Tooltip**
- [x] Should be `background-color: #272D33;`, not white
- [x] Should be `border: none;`
- [x] Remove the label
- [x] Use Filter icon instead of the tooltip icon
- [x] Tooltip content should say "Filter"
- [x] Tooltip is not aligned with top of icon
- [x] Padding should be 4px (.025rem) on all sides
- [x] Should be `font-size: 0.75rem;` (12px)
- [x] Make sure the shark fin is behind the tooltip background
- [x] For example purposes please have the tooltip appear below the icon, not above.

**Comment by aledavila** [comment]
@alisonjoseph 

Ignore the colors. They will be updated once TJs updated colors are merged. The CSS tooltips have the right color but as a hex for the time being. Renaming Default Tooltip to Interactive Tooltip is more complex than simply changing file name. Possible workaround could be name change in website itself. 

Definition: I changed the definition tooltip completely and took made it a separate element instead of creating it with a before. It helped with being able to give it a max width and if more than the max widt
h it will wrap appropriately. And re did the CSS for it. 

Default: As for the JS I added the prevent close when clicking inside the tooltip. Only close when clicking outside. 

Simple --> Icon: Simple tooltip is now Icon only tooltip.

**Review by marijohannessen — APPROVED** [review]
Looks good to me! 🎉 Would it make sense to lose the `v2` part maybe and target this PR towards the `v9` branch instead so we can release it with the new inputs etc?

**Comment by aledavila** [comment]
Yea that sounds good. @marijohannessen would that mean just removing the default and just changing V2 naming back to its default so it’s the default tooltip ? Or Should I pull V9 and make a new branch and just copy the updates onto there ?

**Comment by marijohannessen** [comment]
Yeah, just remove `v2` from all classes etc. And then you can just point this PR to the `v9` branch instead of `master`.

If you click on "Edit" next to the PR title:

screen shot 2018-05-04 at 3 25 08 pm

and then change the base from master to v9:

screen shot 2018-05-04 at 3 25 14 pm

**Review by alisonjoseph — APPROVED** [review]
LGTM! 👍

**Review by asudoh — COMMENTED** [review]
LGTM basically - Great to see this PR getting there @aledavila! Just one little thing - Is it possible for you to remove `yarn-error.log` and `package-lock.json` from the source control before merging? Thanks!

**Comment by alisonjoseph** [comment]
@aledavila since the demo template changes have been merged in to the v9 branch, would be nice to update this from html to handlebar templates before we merge it in.

**Comment by asudoh** [comment]
^ Yes it'll be nice - Simple rename from `.html` to `.hbs` will work for now, but great to merge several component variant's markup using the power of Handlebars. Feel free to ping me if you have any questions - Thanks!

**Review by alisonjoseph — CHANGES_REQUESTED** [review]
Would be good to add correct aria roles.  Ref https://github.com/carbon-design-system/carbon-components/issues/771

https://www.w3.org/TR/wai-aria-practices/#tooltip

>WAI-ARIA Roles, States, and Properties
The element that serves as the tooltip container has role tooltip.
>The element that triggers the tooltip references the tooltip element with aria-describedby.

File: src/components/tooltip/_tooltip.scss [hunk 4]
```diff
@@ -20,43 +20,64 @@
...
     @include reset;
     position: absolute;
     display: none;
     max-width: rem(240px);
     background: $ui-01;
-    padding: $spacing-lg;
+    margin-top: $spacing-xs;
+    padding: $spacing-md;
     border: 1px solid $ui-03;
+    border-radius: 4px;
```
Comment (tw15egan) [inline src/components/tooltip/_tooltip.scss]: Can we switch this to `rem(4px)`?

File: src/components/tooltip/_tooltip.scss [hunk 5]
```diff
@@ -112,20 +133,99 @@
...
+    @include layer('overlay');
+    position: absolute;
+    z-index: 1;
+    display: none;
+    background: $inverse-02;
+    max-width: rem(176px);
+    margin-top: $spacing-sm;
+    padding: $spacing-xs;
+    color: $inverse-01;
+    border-radius: 4px;
```
Comment (tw15egan) [inline src/components/tooltip/_tooltip.scss]: Can we switch this to `rem(4px)`?

File: src/components/tooltip/_tooltip.scss [hunk 6]
```diff
@@ -112,20 +133,99 @@
...
+    cursor: pointer;
+
+    &:after {
+      background: $inverse-02;
+      position: absolute;
+      right: 0;
+      left: 0;
+      width: 0.6rem;
+      height: 0.6rem;
+      margin-left: 24px;
```
Comment (tw15egan) [inline src/components/tooltip/_tooltip.scss]: Can we switch this to `rem(24px)`?

File: src/components/tooltip/_tooltip.scss [hunk 7]
```diff
@@ -134,17 +234,18 @@
...
+      @include typescale('omega');
       @include layer('overlay');
-      max-width: rem(240px);
-      padding: $spacing-lg;
-      border: 1px solid $ui-03;
+      max-width: rem(176px);
       margin-left: 50%;
-      color: $text-01;
+      padding: $spacing-2xs;
+      border-radius: 4px;
```
Comment (tw15egan) [inline src/components/tooltip/_tooltip.scss:244]: Same as above

**Review by tw15egan — APPROVED** [review]
Just a few tiny changes, but other than that it looks great!

**Comment by aledavila** [comment]
WIP: working on accessibility

File: src/components/tooltip/tooltip.hbs [hunk 8]
```diff
@@ -1,10 +1,18 @@
-<a class="bx--tooltip__trigger">
-  Tooltip label
-  <svg tabindex="0" data-tooltip-trigger data-tooltip-target="#unique-tooltip" width="16" height="16" viewBox="0 0 16 16" fill-rule="evenodd">
-    <path d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm0 4c.6 0 1 .4 1 1s-.4 1-1 1-1-.4-1-1 .4-1 1-1zm2 8H6v-1h1V8H6V7h3v4h1v1z"></path>
-  </svg>
-</a>
+<div class="bx--tooltip__label">
```
Comment (alisonjoseph) [inline src/components/tooltip/tooltip.hbs]: We should add the aria-describedby here `<div class="bx--tooltip__label" aria-describedby="unique-tooltip">`

File: src/components/tooltip/tooltip.hbs [hunk 9]
```diff
@@ -1,10 +1,18 @@
...
+  Tooltip Label
+  <div tabindex="0" data-tooltip-trigger data-tooltip-target="#unique-tooltip" role="tooltip" aria-labelledby="unique-tooltip" class="bx--tooltip__trigger">
```
Comment (alisonjoseph) [inline src/components/tooltip/tooltip.hbs]: you can remove `aria-labelledby="unique-tooltip"` here

**Review by alisonjoseph — APPROVED** [review]
Looks good! 🎉

**Pull request merged**